content/6_Advanced/LCT.mdx

```diff
@@ -30,18 +31,309 @@ frequency: 1
+		return s;
+	}
+	friend ostream &operator<<(ostream &s, const mint &v) {
+		s << v.v;
+		return s;
+	}
+};
+// EndCodeSnip
+
+struct Line {
+	mint w = 1, b = 0;
+	mint operator()(mint x) { return w * x + b; }
+	Line operator()(Line f) { return Line{w * f.w, w * f.b + b}; }
+	operator bool() const { return w.v != 1 || b.v != 0; }
+};
+
+const int N = 5e5;
+struct Node;
+typedef struct Node *T;
+struct Node {
+	mint v, s;
+	int sz;
+	inline static int n = 0;
+	// Store Nodes in contiguous array rather than allocating with `new` to
+	// prevent MLE
+	static Node t[2 * N];
+	// c[0] -> left child; c[1] -> right child
+	array<T, 2> c{};
+	T p{};
+	Line f{};
+	bool rev = false;
+
+	Node() {}
+	Node(int v) : v(v), s(v), sz(1), rev(false) {}
+
+	static T init(int v, T l = nullptr, T r = nullptr) {
+		t[n] = Node(v);
+		t[n].attach(0, l);
+		t[n].attach(1, r);
+		t[n].pull();
+		return &t[n++];
+	}
+
+	static int size(T x) { return x ? x->sz : 0; }
+
+	static mint sum(T x) { return x ? x->s : 0; }
+
+	void push() {
+		if (rev) {
+			swap(c[0], c[1]);
+			for (T x : c)
+				if (x) { x->rev ^= rev; }
+			rev = false;
+		}
+		if (f) {
+			v = f(v);
+			s = f.w * s + f.b * sz;
+			for (T x : c)
+				if (x) { x->f = f(x->f); }
+			f = Line{1, 0};
+		}
+	}
+
+	void pull() {
+		for (T x : c)
+			if (x) { x->push(); }
+		sz = size(c[0]) + size(c[1]) + 1;
+		s = sum(c[0]) + sum(c[1]) + v;
+	}
+
+	bool side() { return p->c[1] == this; }
+
+	void attach(int i, T x) {
+		if (x) { x->p = this; }
+		c[i] = x;
+		push();
+		pull();
+	}
+	T detach(int i) {
+		push();
+		T t = c[i];
+		if (c[i]) { c[i]->p = nullptr; }
+		c[i] = nullptr;
+		pull();
+		return t;
+	}
+
+	void rotate() {
+		T p = this->p;  // parent
+		T g = p->p;     // grandparent
+		p->push();
+		push();
+		int i = side();
+		if (g) {
+			g->attach(p->side(), this);
+		} else this->p = nullptr;
+		p->attach(i, c[!i]);
+		attach(!i, p);
+	}
+
+	void splay() {
+		for (; p; rotate()) {
+			if (p->p) {
+				if (side() == p->side()) {
+					p->rotate();
+				}  // zig-zig
+				else {
+					rotate();
+				}  // zig-zag
+			}
+		}
+	}
+
+	T find(int i) {
+		push();
+		if (i < size(c[0])) {
+			return c[0]->find(i);
+		} else if (i > size(c[0])) {
+			return c[1]->find(i - size(c[0]) - 1);
+		} else {
+			return this;
+		}
+	}
+
+	friend ostream &operator<<(ostream &os, T s) {
+		if (!s) return os;
+		os << s->c[0];
+		os << s->v << " ";
+		return os << s->c[1];
+	}
+};
+
+Node Node::t[2 * N];
+
+struct SplayTree {
+	T root = nullptr;
+	T splay(int x) {
+		T t = root->find(x);
+		t->splay();
+		return root = t;
+	}
+
+	pair<SplayTree, SplayTree> split(int i) {  // splits into {[0, i), [i, )}
+		if (i == (root ? root->sz : 0)) {
+			return make_pair(SplayTree{root}, SplayTree{nullptr});
+		}
+		T t = splay(i)->detach(0);
+		return make_pair(SplayTree{t}, SplayTree{root});
+	}
+
+	SplayTree join(SplayTree o) {
+		if (!root) { return root = o.root, *this; }
+		splay(root->sz - 1);
+		root->attach(1, o.root);
```

Comment:
yeah same thing about these splay tree funcs

Reply:
for the splaytree funcs yes they should probably be public, I can private some of the node internal stuff though